src/Symfony/Component/DependencyInjection/ServiceLocator.php

```diff
@@ -45,23 +59,56 @@ public function has($id)
     public function get($id)
     {
         if (!isset($this->factories[$id])) {
-            throw new ServiceNotFoundException($id, null, null, array_keys($this->factories));
+            throw new ServiceNotFoundException($id, end($this->loading) ?: null, null, array(), $this->createServiceNotFoundMessage($id));
         }
 
-        if (true === $factory = $this->factories[$id]) {
-            throw new ServiceCircularReferenceException($id, array($id, $id));
+        if (isset($this->loading[$id])) {
+            $ids = array_values($this->loading);
+            $ids = array_slice($this->loading, array_search($id, $ids));
+            $ids[] = $id;
+
+            throw new ServiceCircularReferenceException($id, $ids);
         }
 
-        $this->factories[$id] = true;
+        $this->loading[$id] = $id;
         try {
-            return $factory();
+            return $this->factories[$id]();
         } finally {
-            $this->factories[$id] = $factory;
+            unset($this->loading[$id]);
         }
     }
 
     public function __invoke($id)
     {
         return isset($this->factories[$id]) ? $this->get($id) : null;
     }
+
+    private function createServiceNotFoundMessage($id)
+    {
+        if ($this->loading) {
+            $msg = sprintf('The service "%s" has a dependency on a non-existent service "%s".', end($this->loading), $id);
+            $msg .= sprintf(' This locator only knows about "%s"', implode('", "', array_keys($this->factories)));
```

Comment:
How does this situation happen? Once I use `$this->container->get('foo')`, if `foo` depends on a non-existent `bar` service, wouldn't the container have already exploded at build time?

Reply:
see new test case :)